### ISSUE TITLE
Enabling openTracing with no tracer class configured impacts performance

### ISSUE THREAD

**Issue by gjdeval** [OpenLiberty/open-liberty#12613]
A performance issue occurs when openTracing is configured in the server.xml (individually or as part of a microProfile feature bundle) and no tracer class is configured. What seems to happen is that OpentracingTracerManager keeps looking for a tracer, on every request. This can have a large impact on the throughput of the server if the workload is mostly HTTP requests/responses with little service logic in the app.

We discovered this issue in 20006, but the problem is not new - it can be seen at least as far back as 19003.

**Comment by fmhwong** [comment]
@gjdeval Do you have numbers comparing with and without `mpOpenTracing-1.3`?  Previously, the app name looked was identified slow and we made improvement there.

**Comment by jdmcclur** [comment]
@fmhwong -  I think we fixed a problem when a tracer impl is provided. In this case, no tracer impl is provided and it seems to build some sort of jaeger tracer on every call? There is some overhead in 19.0.0.3, but it's a lot worse recently (maybe since when we shipped jaeger support? I am not sure).

**Comment by gjdeval** [comment]
Joe profiled this with one test app and saw ~78% of time being spent in/under
```
OpentracingTracerManager.ensureTracer(Ljava/lang/String;)Lio/opentracing/Tracer;
```
I ran a similar experiment with a simple ping app and saw throughput drop from ~45K baseline to less than 1K tps with mpOpenTracing-1.3 added to the server.xml. When I profiled this, it showed high lock contention under the same 'ensureTracer' method.

Seems that Liberty should be looking for the tracer class preferably only once, on server start, or when there is a server config change maybe ... certainly not on every request, as appears to be the case now.

**Comment by fmhwong** [comment]
This PR https://github.com/OpenLiberty/open-liberty/pull/11892 got merged on 4/23/2020 and I suppose it is included in 20.0.0.5.  Not sure this is making the lock contention issue better or worse.

**Comment by gjdeval** [comment]
PR #11892 seems to have side-stepped the actual issue, perhaps. Performance with my test app is the same in 20003 as in 20006. 

The basic issue is that Liberty keeps doing work _on every single HTTP request_ to try to get a non-existent tracer class. 

The first time through, when we figure out that there is no tracer configured for this app, we need to save that information somehow, so that for subsequent HTTP requests we do not spend time trying to get a non-existent tracer.

**Comment by NottyCode** [comment]
I totally support improving the performance here, but not in a way that would mean someone updating the configuration to add the tracer implementation would need to restart the server.

I looked at the code and there is currently a classloader leak in the OpenTracerManager. An application restart (common in development) will not trigger a new tracer to be created. The map will have the prior one which means that the classloader for the first app start will never be cleaned up. This isn't as bad as if we had one instance per restart, but we should get it fixed. In addition to not leaking a classloader if an application restarts you want to ensure you get the most recent code.

**Comment by fmhwong** [comment]
Fixed mpOT-1.3 with PR #12763
